docs/tech-notes/range-merges.md

```diff
@@ -0,0 +1,259 @@
+admin commands, DistSender will automatically route the request to the
+leaseholder of the range, but there is no guarantee that the store will retain
+its lease while the admin command is executing.
+
+Note that an AdminMerge request takes no arguments, as there is no choice in
+what range will be merged. The recipient of the AdminMerge will always be the
+LHS, subsuming range, and its right neighbor at the moment that the AdminMerge
+command begins executing will always be the RHS, subsumed range.
+
+It would perhaps have been wise to require that the AdminMerge specify the
+descriptor of the intented subsumption target. It is, at present, possible for
+someone to send an admin merge to P, intending to merge P's right neighbor Q,
+but Q can split in the meantime and so instead P merges with a different range.
+Luckily, if this proves to be a problem, it is an easy extension to the API.
+
+It would have been reasonable to have instead used the RHS to coordinate the
+merge. That is, the RHS would have been the subsuming range, and the LHS would
+have been the subsumed range. Using the LHS is slightly easier, however, due to
+an oddity of key encoding and range bounds. It is trivial for a range to send a
+request to its right neighbor, as it simply addresses the request to its end
+key, but difficult to send a request to its left neighbor.
+
+At the time of writing, only the merge queue initiates merges, and it does so
+by bypassing DistSender and invoking the AdminMerge command directly on the
+local replica. At some point in the future, we may wish to expose manual merges
+via SQL, at which point the SQL layer will need to send proper AdminMerge
+requests through the KV API.
+
+## Merge transaction
+
+The merge transaction piggybacks on CockroachDB's strict serializability to
+provide much of the necessary synchronization for the bookkeeping updates. For
+example, merges cannot occur concurrently with any splits or replica changes on
+the implicated ranges, because the merge transaction will naturally conflict
+with those split transaction and change replicas transactions, as both
+transactions will attempt to write updated range descriptors and conflict. No
+additional code was needed to enforce this, as our standard transaction
+conflict detection mechanisms kick in here (write intents, the timestamp cache,
+the command queue, etc.).
+
+There was, however, one knotty edge case that was not immediately eliminated by
+transaction serializability. Suppose we have our standard misaligned replica
+set situation:
+
+```
+Store 1    Store 2     Store 3     Store 4
++-----+    +-----+     +-----+     +-----+
+| P Q |    | P Q |     |     |     | P Q |
++-----+    +-----+     +-----+     +-----+
+```
+
+In an unfortunate twist of fate, a rebalance of P from store 2 to store 3
+begins at the same time as a merge of P and Q begins. Let's quickly cover the
+valid outcomes of this race.
```

Comment:
These are the valid outcomes, but are they all possible? Are there any important mechanisms in place ensuring no other outcomes occur?

Reply:
Yes, they're all possible. Perhaps this section needs a bit more love. The idea is that either the rebalance commits first or the merge commits first. (One of those things must happen. I guess there are cases where the rebalance aborts or the merge aborts for unrelated reasons, but those don't matter.) Once you assume that the rebalance commits first, then the merge must abort. If you instead assume that the merge commits first, the rebalance must abort—but this case is broken into two because there the rationale for why the abort happens depends on precisely when the merge commit occurs.